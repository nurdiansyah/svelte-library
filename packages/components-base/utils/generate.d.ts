interface Options {
    size?: number;
    prefix?: string;
}
export declare const getId: ({ prefix, size }: Options) => string;
export {};
