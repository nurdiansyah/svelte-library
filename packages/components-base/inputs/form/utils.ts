// Supports determination of isControlled().
// Controlled input accepts its current value as a prop.
//
// @param value
// @returns {boolean} true if string (including '') or number (including zero)
export function hasValue(value) {
  return value != null && !(Array.isArray(value) && value.length === 0);
}

// Determine if field is empty or filled.
// Response determines if label is presented above field or as placeholder.
//
// @param value
// @param SSR
// @returns {boolean} False when not present or empty string.
//                    True when any number or string with length.
export function isFilled(value: any, SSR = false) {
  return hasValue(value) && value !== "";
}
