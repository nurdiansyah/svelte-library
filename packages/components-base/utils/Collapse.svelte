<script lang="ts">
  import { clsx } from "./clsx";

  import { createEventDispatcher } from "svelte";
  import { slide } from "svelte/transition";
  const noop = () => undefined;

  export let show: boolean = false;
  let { class: className } = $$props;
  let classes: string | undefined;
  export let onEntering: (e?: Event) => any = noop;
  export let onEntered: (e?: Event) => any = noop;
  export let onExiting: (e?: Event) => any = noop;
  export let onExited: (e?: Event) => any = noop;

  $: classes = clsx(className, "dbx-collapse", show && "-show");
</script>

<div
  {...$$restProps}
  class={classes}
  transition:slide|local
  on:introstart
  on:introend
  on:outrostart
  on:outroend
  on:introstart={onEntering}
  on:introend={onEntered}
  on:outrostart={onExiting}
  on:outroend={onExited}>
  <slot />
</div>

<style lang="scss" global>
  .dbx-collapse:not(.-show) {
    display: none;
  }
</style>
