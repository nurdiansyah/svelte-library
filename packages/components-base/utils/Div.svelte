<script>
  export let node = undefined;
</script>

<div {...$$props} bind:this={node}>
  <slot />
</div>
