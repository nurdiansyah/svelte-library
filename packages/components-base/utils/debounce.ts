// Corresponds to 10 frames at 60 Hz.
// A few bytes payload overhead when lodash/debounce is ~3 kB and debounce ~300 B.
export function debounce(func: (arg0: any, arg1: any[]) => void, wait = 166) {
  let timeout: NodeJS.Timeout;
  function debounced(...args: any[]) {
    // eslint-disable-next-line consistent-this
    const that = this;
    const later = () => {
      func.apply(that, args);
    };
    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
  }

  debounced.clear = () => {
    clearTimeout(timeout);
  };

  return debounced;
}
