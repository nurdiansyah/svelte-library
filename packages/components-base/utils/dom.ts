export const ownerDocument = (node: Node) => (node && node.ownerDocument) || document;
export const ownerWindow = (node: Node) => {
  const doc = ownerDocument(node);
  return doc.defaultView || window;
};

// A change of the browser zoom change the scrollbar size.
// Credit https://github.com/twbs/bootstrap/blob/3ffe3a5d82f6f561b82ff78d82b32a7d14aed558/js/src/modal.js#L512-L519
export const getScrollbarSize = () => {
  const scrollDiv = document.createElement("div");
  scrollDiv.style.width = "99px";
  scrollDiv.style.height = "99px";
  scrollDiv.style.position = "absolute";
  scrollDiv.style.top = "-9999px";
  scrollDiv.style.overflow = "scroll";

  document.body.appendChild(scrollDiv);
  const scrollbarSize = scrollDiv.offsetWidth - scrollDiv.clientWidth;
  document.body.removeChild(scrollDiv);

  return scrollbarSize;
};

export const getOriginalBodyPadding = () => {
  const style = window ? window.getComputedStyle(document.body, null) : { getPropertyValue: (e?: any) => undefined };

  return parseInt((style && style.getPropertyValue("padding-right")) || "0", 10);
};

export const setScrollbarWidth = (padding) => {
  if (padding > 0) {
    document.body.style.paddingRight = `${padding}px`;
  }
};

export const isBodyOverflowing = () => (window ? document.body.clientWidth < window.innerWidth : false);
