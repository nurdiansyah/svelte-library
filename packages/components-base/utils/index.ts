export { default as Collapse } from "./Collapse.svelte";
export { default as Scrollbar } from "./Scrollbar.svelte";
export * as actions from "./actions";
export * from "./generate";
export * from "./style";
export * from "./popper";
