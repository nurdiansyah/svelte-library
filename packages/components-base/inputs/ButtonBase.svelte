<script lang="ts">
  import type { SvelteComponent } from "svelte";

  import { clsx } from "../utils";
  import { ripple, focusVisible as focusVisibleAction } from "../actions";

  let { class: className, href } = $$props;
  let classes: string;
  let focusVisible: boolean = false;

  export let disabled: boolean = false;
  export let link: boolean = false;
  export let focusVisibleClass: string | undefined = undefined;
  // TODO ripple not implement
  export let rippleOptions = {};
  export let focusRippleEnable: boolean = true;
  export let tagDiv: boolean = false;

  let isFocusVisible: (e: Event) => undefined;
  let onBlurVisible: (e: Event) => undefined;

  $: classes = clsx(className, "button-base", focusVisible && !disabled && (focusVisibleClass || "-focus-visible"));

  const handleIsFocusVisible = ({ isFocusVisible: _isFocusVisible, onBlurVisible: _onBlurVisible }) => {
    isFocusVisible = _isFocusVisible;
    onBlurVisible = _onBlurVisible;
  };

  const handleBlur = (e: Event) => {
    focusVisible = false;
    onBlurVisible && onBlurVisible(e);
  };

  const handleFocus = (e: Event) => {
    if (isFocusVisible && isFocusVisible(e)) {
      focusVisible = true;
    }
  };
</script>

{#if href || link}
  <a
    {...$$restProps}
    class={classes}
    on:click
    on:blur
    on:focus
    on:blur={handleBlur}
    on:focus={handleFocus}
    href={href || '#'}
    use:focusVisibleAction={{ onFocusVisible: handleIsFocusVisible }}>
    <slot />
  </a>
{:else if tagDiv}
  <div
    {...$$restProps}
    class={classes}
    on:click
    on:blur
    on:focus
    on:blur={handleBlur}
    on:focus={handleFocus}
    use:focusVisibleAction={{ onFocusVisible: handleIsFocusVisible }}>
    <slot />
  </div>
{:else}
  <button
    {...$$restProps}
    class={classes}
    on:click
    on:blur
    on:focus
    on:blur={handleBlur}
    on:focus={handleFocus}
    use:focusVisibleAction={{ onFocusVisible: handleIsFocusVisible }}>
    <slot />
  </button>
{/if}
{#if focusRippleEnable}
  <div use:ripple={rippleOptions} />
{/if}

<style src="./styles/button-base.scss" global>
</style>
