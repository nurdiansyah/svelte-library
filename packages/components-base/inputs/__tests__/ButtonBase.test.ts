import { itConformance } from "../../utils/test-utils";
import ButtonBase from "../ButtonBase.svelte";

describe("ButtonBase", () => {
  const testId = "button-base";
  const props: any = { "data-testid": testId };

  const render = itConformance(ButtonBase, { props });

  it("check: href || link", () => {
    const { container } = render({ link: true });
    expect(container).toMatchSnapshot();
  });

  it("check: div", () => {
    const { getByTestId } = render({ tagDiv: true });
    expect(getByTestId(testId).tagName).toBe("DIV");
  });

  describe("directive focusVisible", () => {
    it("check: ", () => {});
  });
});
