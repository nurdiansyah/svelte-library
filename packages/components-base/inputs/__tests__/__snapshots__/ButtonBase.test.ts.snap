// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonBase cek render komponen tanpa error dan snapshot 1`] = `
<body>
  <div>
    <button
      class="button-base"
      data-testid="button-base"
    />
     
    <div
      class="ripple-container"
    />
    
  </div>
</body>
`;

exports[`ButtonBase check: href || link 1`] = `
<body>
  <div>
    <a
      class="button-base"
      data-testid="button-base"
      href="#"
    />
     
    <div
      class="ripple-container"
    />
    
  </div>
</body>
`;
